vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,314 @@
+     * Defaults to "manifest.yml" if not specified otherwise.
+     */
+    'manifestFile',
+    /**
+     * The {@code String} path of the Yaml file to produce as output.
+     * If not specified this will default to {@code manifestFile} and overwrite it.
+     */
+    'outputManifestFile',
+    /**
+     * The {@code String} path of the Yaml file containing the variables' values to use as a replace in the manifest file. <br>
+     * Defaults to "manifest-variables.yml" if not specified otherwise
+     */
+    'variablesFile'
+]
+
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+/**
+ * Step that substitutes variables in a given YAML file with those specified in a another. The format to reference a variable
+ * in the YAML file is to use double parentheses {@code ((} and {@code ))}, e.g. {@code ((variableName))}.<br>
+ * A declaration of a variable and assignment of its value is simply done as a property in the variables YAML file.
+ * <p>
+ * The format follows <a href="https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#variable-substitution">Cloud Foundry standards</a>.
+ * <p>
+ * The step is activated by the presence of both a {@code manifest.yml} and a variables file. Names of both files are configurable.
+ * <p>
+ * Usage: yamlSubstituteVariables manifestFile: "manifest.yml", variablesFile: "manifest-variables.yml"
+ *
+ * @param arguments - the map of arguments.
+ */
+@GenerateDocumentation
+void call(Map<String, String> arguments) {
+    handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: arguments) {
+        def script = checkScript(this, arguments)  ?: this
+
+        // load default & individual configuration
+        Map config = ConfigurationHelper.newInstance(this)
+                                        .loadStepDefaults()
+                                        .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+                                        .mixinStageConfig(script.commonPipelineEnvironment, arguments.stageName ?: env.STAGE_NAME, STEP_CONFIG_KEYS)
+                                        .mixin(arguments, PARAMETER_KEYS)
+                                        .use()
+
+        String manifestFilePath = config.manifestFile ?: "manifest.yml"
+        String variablesFilePath = config.variablesFile ?: "manifest-variables.yml"
+        String outputFilePath = config.outputManifestFile ?: manifestFilePath
+
+        // Parameter for tests only. No public API.
+        // This is to skip manifest file deletion in tests, which
+        // would be a nuisance, as it alters the sources repository.
+        Boolean skipDeletion = arguments.skipDeletion ?: false
+
+        File manifestFile = new File(manifestFilePath)
+        File variablesFile = new File(variablesFilePath)
+
+        if (!manifestFile.exists()) {
+            echo "[YamlSubstituteVariables] Could not find YAML file at ${manifestFilePath}. Skipping variable substitution."
```

Comment:
Not sure if simply returning in case the file does not exist is a good pattern. With such an approach we silently skip the attempt to replace content. Wouldn't it be better to fail with an exception? I mean, the variable substitution step is introduced by a pipeline author for a reason. In case a certain file does not exist we should fail here rather than failing later.

Reply:
This was discussed with Kai and Ralph and I understood, we wanted a behaviour that (auto-magically) activates itself, when it should and step out of the way if it is not applicable.
That's way it is now coded that way. I understood a lot of components in Piper auto-trigger if certain conditions are met, and step aside if that is not the case.

Reply:
Meta: I have a name clash with the names above (knowing at least more than one "Ralf". Maybe using the github names @xyz might be better.

In my opinion things which happens "automagically" makes support less easy. In case I want something to happen I use the step in my pipeline script. In case I don't want something to happen I don't. Now it is clear to the poor guy who has to provide support what is going on. In case it depends (... should be a rare use case, I basically think "it depends" might be some smell) I vote for checking the details outside the step and call the step only in case the check succeeds: e.g.

```
if(checkFoo()) {
    doFoo script: this
} 
```

rather than calling `doFoo` always and making the decision inside.

But I'm not the king and piper-library-os is not my realm. In case you think that behaviour is appropriate for the use-case: go ahead.

Reply:
I will follow up on this question with the named colleagues. Unfortunately they do not show up as Github users here (so I could not mention them the "GitHub" way).

Reply:
I checked this again with the colleagues today, and I understood, that we will keep the behaviour as implemented, i.e.:

* Ability to specify the name of a manifest to replace variables in, default: `manifest.yml`
* Ability to specify the name of a variables Yaml to resolve variables from, default: `manifest-variables.yml` 
* If both files exist, the step will activate automatically, otherwise it will not perform any substitution but print warnings.

This should yield a clean call in a pipeline and resort to a meaningful default behaviour based on convention.
That said, I can't say I am all for "auto-magic" myself. I usually also prefer more explicit control over what's happening. There is always a trade-off with convenience.

Note also: the script that shows above behaviour is now named `cfManifestSubstituteVariables` and is more tailored towards manifest files.
The substitution based on Yaml objects is outsourced now and available as its own script (named `yamlSubstituteVariables` as before).